apps/module-presale/contracts/Presale.sol

```diff
@@ -0,0 +1,366 @@
+    * @param _buyer address The buyer address to refund.
+    * @param _vestedPurchaseId uint256 The purchase id to refund.
+    */
+    function refund(address _buyer, uint256 _vestedPurchaseId) public {
+        require(currentSaleState() == SaleState.Refunding, ERROR_INVALID_STATE);
+
+        // Recall how much contribution tokens to refund for this purchase.
+        uint256 tokensToRefund = purchases[_buyer][_vestedPurchaseId];
+        require(tokensToRefund > 0, ERROR_NOTHING_TO_REFUND);
+        purchases[_buyer][_vestedPurchaseId] = 0;
+
+        // (presale) ~~~> contribution tokens ~~~> (buyer)
+        require(contributionToken.transfer(_buyer, tokensToRefund), ERROR_TOKEN_TRANSFER_REVERTED);
+
+        // (buyer) ~~~> project tokens ~~~> (Token manager)
+        // Note: this assumes that the buyer didn't transfer any of the vested tokens.
+        // The assumption can be made, considering the imposed restriction of fundingPeriod < vestingCliffPeriod < vestingCompletePeriod.
+        (uint256 tokensSold,,,,) = projectTokenManager.getVesting(_buyer, _vestedPurchaseId);
+        projectTokenManager.revokeVesting(_buyer, _vestedPurchaseId);
+
+        // (Token manager) ~~~> project tokens ~~~> (burn 💥)
+        projectTokenManager.burn(address(projectTokenManager), tokensSold);
+
+        emit TokensRefunded(_buyer, tokensToRefund, tokensSold, _vestedPurchaseId);
+    }
+
+    /**
+    * @notice Closes a sale that has reached the funding goal, sending raised funds to the fundraising pool and the beneficiary address, and initializes the Fundraising app by adding the raised funds as collateral tokens.
+    */
+    function close() public {
+        require(currentSaleState() == SaleState.GoalReached, ERROR_INVALID_STATE);
+
+        // (presale) ~~~> contribution tokens ~~~> (beneficiary)
+        uint256 tokensForBeneficiary = totalRaised.mul(percentFundingForBeneficiary).div(PPM);
+        require(contributionToken.transfer(beneficiaryAddress, tokensForBeneficiary), ERROR_TOKEN_TRANSFER_REVERTED);
+
+        // (presale) ~~~> contribution tokens ~~~> (pool)
+        uint256 tokensForPool = contributionToken.balanceOf(address(this));
+        require(contributionToken.transfer(fundraisingPool, tokensForPool), ERROR_TOKEN_TRANSFER_REVERTED);
+
+        saleClosed = true;
+
+        emit SaleClosed();
+    }
+
+    /*
+     * Getters
+     */
+
+    /**
+    * @notice Calculates the number of project tokens that would be obtained for `_amount` contribution tokens.
+    * @param _amount uint256 The amount of contribution tokens to be converted into project tokens.
+    */
+    function contributionToProjectTokens(uint256 _amount) public view returns (uint256) {
+        return _amount.mul(tokenExchangeRate);
+    }
+
+    /**
+    * @notice Calculates the current state of the sale.
+    */
+    function currentSaleState() public view returns (SaleState) {
```

Comment:
This is the function we should update to use an automatic starting date, right ?

Reply:
The automatic start date is already implemented. If you specify `startDate != 0` upon initialization, the Presale will automatically open at the specified date.